arduino-ide-extension/src/browser/boards/boards-service-provider.ts

```diff
@@ -41,402 +141,327 @@ export class BoardsServiceProvider
-    oldState: Port[];
-  }>();
-  private readonly inheritedConfig = new Deferred<BoardsConfig.Config>();
-
+  @inject(BoardsDataStore)
+  private readonly boardsDataStore: BoardsDataStore;
+  @optional()
+  @inject(BoardListDumper)
+  private readonly boardListDumper?: BoardListDumper;
+
+  private _boardsConfig = emptyBoardsConfig();
+  private _detectedPorts: DetectedPorts = {};
+  private _boardList = this.createBoardListUI(createBoardList({}));
+  private _boardListHistory: Mutable<BoardListHistory> = {};
+  private _ready = new Deferred<void>();
+
+  private readonly boardsConfigDidChangeEmitter =
+    new Emitter<BoardsConfigChangeEvent>();
+  readonly onBoardsConfigDidChange = this.boardsConfigDidChangeEmitter.event;
+
+  private readonly boardListDidChangeEmitter = new Emitter<BoardListUI>();
   /**
-   * Used for the auto-reconnecting. Sometimes, the attached board gets disconnected after uploading something to it.
-   * It happens with certain boards on Windows. For example, the `MKR1000` boards is selected on post `COM5` on Windows,
-   * perform an upload, the board automatically disconnects and reconnects, but on another port, `COM10`.
-   * We have to listen on such changes and auto-reconnect the same board on another port.
-   * See: https://arduino.slack.com/archives/CJJHJCJSJ/p1568645417013000?thread_ts=1568640504.009400&cid=CJJHJCJSJ
+   * Emits an event on board config (port or board) change, and when the discovery (`board list --watch`) detected any changes.
    */
-  protected latestValidBoardsConfig:
-    | RecursiveRequired<BoardsConfig.Config>
-    | undefined = undefined;
-  protected latestBoardsConfig: BoardsConfig.Config | undefined = undefined;
-  protected _boardsConfig: BoardsConfig.Config = {};
-  protected _attachedBoards: Board[] = []; // This does not contain the `Unknown` boards. They're visible from the available ports only.
-  protected _availablePorts: Port[] = [];
-  protected _availableBoards: AvailableBoard[] = [];
-
-  private lastBoardsConfigOnUpload: BoardsConfig.Config | undefined;
-  private lastAvailablePortsOnUpload: Port[] | undefined;
-  private boardConfigToAutoSelect: BoardsConfig.Config | undefined;
-
-  /**
-   * Unlike `onAttachedBoardsChanged` this event fires when the user modifies the selected board in the IDE.\
-   * This event also fires, when the boards package was not available for the currently selected board,
-   * and the user installs the board package. Note: installing a board package will set the `fqbn` of the
-   * currently selected board.
-   *
-   * This event is also emitted when the board package for the currently selected board was uninstalled.
-   */
-  readonly onBoardsConfigChanged = this.onBoardsConfigChangedEmitter.event;
-  readonly onAvailableBoardsChanged =
-    this.onAvailableBoardsChangedEmitter.event;
-  readonly onAvailablePortsChanged = this.onAvailablePortsChangedEmitter.event;
-
-  private readonly _reconciled = new Deferred<void>();
```

Comment:
Very happy we are getting rid of this deferred promise....!

Reply:
> What's the problem with the deferred promise?

Sorry for the ambiguity, I was in fact referring to the renaming :), `reconciled` confused me, "getting rid" is not really the correct term....